pilot/pkg/model/extensions.go

```diff
@@ -56,6 +74,58 @@ type WasmPluginWrapper struct {
 	WasmExtensionConfig *envoyWasmFilterV3.Wasm
 }
 
+func (p *WasmPluginWrapper) MatchListener(proxyLabels map[string]string, li WasmPluginListenerInfo) bool {
+	workloadMatch := (p.Selector == nil || labels.Instance(p.Selector.MatchLabels).SubsetOf(proxyLabels))
+	return workloadMatch && matchTrafficSelectors(p.Match, li)
+}
+
+type WasmPluginListenerInfo struct {
+	Port  int
+	Class istionetworking.ListenerClass
+}
+
+// If anyListener is used as a listener info,
+// the listener is matched with any TrafficSelector.
+var anyListener = WasmPluginListenerInfo{
+	Port:  0,
+	Class: istionetworking.ListenerClassUndefined,
+}
+
+func matchTrafficSelectors(ts []*extensions.WasmPlugin_TrafficSelector, li WasmPluginListenerInfo) bool {
+	if (li.Class == istionetworking.ListenerClassUndefined && li.Port == 0) || len(ts) == 0 {
+		return true
+	}
+
+	for _, match := range ts {
+		if matchMode(match.Mode, li.Class) && matchPorts(match.Ports, li.Port) {
+			return true
+		}
+	}
+	return false
+}
+
+func matchMode(workloadMode typeapi.WorkloadMode, class istionetworking.ListenerClass) bool {
+	switch workloadMode {
+	case typeapi.WorkloadMode_CLIENT_AND_SERVER, typeapi.WorkloadMode_UNDEFINED:
+		return true
+	default:
+		return workloadMode == workloadModeForListenerClass(class)
+	}
+}
+
+func matchPorts(portSelectors []*typeapi.PortSelector, port int) bool {
+	if len(portSelectors) == 0 {
+		// If there is no specified port, match with all the ports.
+		return true
+	}
+	for _, ps := range portSelectors {
+		if ps.GetNumber() != 0 && ps.GetNumber() == uint32(port) {
```

Comment:
very nit: slightly better to check `ps` against `nil` than GetNumber 0 since it could be explicitly set to `0`. In practice this is impossible due to validation, though